compose/config/validation.py

```diff
@@ -107,15 +113,60 @@ def _parse_key_from_error_msg(error):
     def _clean_error_message(message):
         return message.replace("u'", "'")
 
-    def _parse_valid_types_from_schema(schema):
+    def _parse_valid_types_from_validator(validator):
         """
-        Our defined types using $ref in the schema require some extra parsing
-        retrieve a helpful type for error message display.
+        A validator value can be either an array of valid types or a string of
+        a valid type. Parse the valid types and prefix with the correct article.
         """
-        if '$ref' in schema:
-            return schema['$ref'].replace("#/definitions/", "").replace("_", " ")
+        if isinstance(validator, list):
+            if len(validator) >= 2:
+                first_type = anglicize_validator(validator.pop(0))
+                last_type = anglicize_validator(validator.pop())
```

Comment:
I'm not sure if validator is re-used at all, but I would probably err on the side of caution here and not do any mutation.

Something like this would work:

```python
first_type, last_type, validator = validator[0], validator[-1], validator[1:-1]
```